File: KthSmallestEle_BST.py
# Definition for a binary tree node.
# class TreeNode:
    # def __init__(self, val=0, left=None, right=None):
    #     self.val = val
    #     self.left = left
    #     self.right = right
    #     self.countleft=0
class Solution:
    """Recursive approach
    Time complexity-O(n)
    Space complexity-O(1)"""
    def __init__(self):
        self.count=0
        self.result=-1
        
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        
        self.helper(root, k)
        return self.result
    
    def helper(self, root, k):
        if not root:
            return None
        if self.result==-1:
            self.helper(root.left, k)
        self.count+=1
        if self.count==k:
            self.result= root.val
        if self.result==-1:
            self.helper(root.right, k)
        
        
        
        

File: kthSmallestEleBST.java
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
// We do an inorder traversal and keep track of a count var
// Return result when count = k

class Solution {
    int count = 0;
    TreeNode res;
    public int kthSmallest(TreeNode root, int k) {     
        inorder(root, k);
        return res.val;
    }
    private void inorder(TreeNode root, int k){
        if(root == null)
            return;
        if(count<k)
            inorder(root.left,k);
        count++;
        if(count == k) {
            res = new TreeNode(root.val);
        }
        if(count<k)
            inorder(root.right, k);
    }
}